demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_payload_operations.h

```diff
@@ -0,0 +1,102 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* Standard includes. */
+#include <stdlib.h>
+#include <stdbool.h>
+
+/**
+ * @brief Creates the document to be published to the CreateCertificateFromCsr
+ * API in order to request a certificate from AWS IoT for the included
+ * Certificate Signing Request (CSR).
+ *
+ * @param[in] pBuffer Buffer into which to write the publish document.
+ * @param[in] bufferLength Length of #buffer.
+ * @param[in] pCsr The CSR to include in the document.
+ * @param[in] csrLength The length of #pCsr.
+ * @param[out] pOutLengthWritten The length of the publish document.
+ */
+bool generateCsrRequest( char * pBuffer,
+                         size_t bufferLength,
+                         const char * pCsr,
+                         size_t csrLength,
+                         size_t * pOutLengthWritten );
+
+/**
+ * @brief Creates the document to be published to the RegisterThing API in order
+ * to activate the provisioned certificate and receive a Thing name.
+ *
+ * @param[in] pBuffer Buffer into which to write the publish document.
+ * @param[in] bufferLength Length of #buffer.
+ * @param[in] pCertificateOwnershipToken The certificate's certificate ownership
+ * token.
+ * @param[in] certificateOwnershipTokenLength Length of #certificateOwnershipToken.
+ * @param[out] pOutLengthWritten The length of the publish document.
+ */
+bool generateRegisterThingRequest( char * pBuffer,
+                                   size_t bufferLength,
+                                   const char * pCertificateOwnershipToken,
+                                   size_t certificateOwnershipTokenLength,
+                                   const char * pSerial,
+                                   size_t serialLength,
+                                   size_t * pOutLengthWritten );
+
+/**
+ * @brief Extracts the certificate, certificate ID, and certificate ownership
+ * token from a CreateCertificateFromCsr accepted response.
+ *
+ * @param[in] pResponse The response document.
+ * @param[in] length Length of #response.
+ * @param[in] pCertificate The buffer to which to write the certificate.
+ * @param[in,out] pCertificateLength The length of #certificate. The written
+ * length is output here.
+ * @param[in] pCertificateId The buffer to which to write the certificate ID.
+ * @param[in,out] pCertificateIdLength The length of #certificateId. The written
+ * length is output here.
+ * @param[in] pOwnershipToken The buffer to which to write the certificate
+ * ownership token.
+ * @param[in,out] pOwnershipTokenLength The length of #ownershipToken. The written
+ * length is output here.
+ */
+bool parseCsrResponse( const char * pResponse,
+                       size_t length,
+                       char * pCertificate,
```

Comment:
It would be more expressive to use `Buffer` suffix in the output parameters for the extracted fields to make it obvious that the information is copied into the buffers

Reply:
As the certificate is provided in DER format when using CBOR encoding, it will be better to use `uint8_t *` type as well as use `byte_string` functions of tinyCBOR for code semantics